codegen/src/api/errors.rs

```diff
@@ -38,31 +38,63 @@ pub fn generate_has_module_error_impl(
         .ty()
         .type_def();
 
-    // Slightly older versions of substrate have a `DispatchError::Module { index, error }`
-    // variant. Newer versions have something like a `DispatchError::Module (Details)` variant.
-    // We check to see which type of variant we're dealing with based on the metadata, and
-    // generate the correct code to handle either older or newer substrate versions.
-    let module_variant_is_struct = if let TypeDef::Variant(details) = dispatch_error_def {
+    // Different substrate versions will have a `DispatchError::Module` of the following form,
+    // ordered by versions:
+    //
+    // Case 1. DispatchError::Module { index: u8, error: u8 }
+    // Case 2. DispatchError::Module ( sp_runtime::ModuleError { index: u8, error: u8 } )
+    // Case 3. DispatchError::Module ( sp_runtime::ModuleError { index: u8, error: [u8; 4] } )
+    //
+    // To handle all cases and maintain backward compatibility, the type of the variant is inspected
+    // based on the metadata.
+    // If the variant has a named field (i.e, Case 1) the variable `module_variant_is_struct` is
+    // true. If the error is of form `u8`, then `module_legacy_err` is True.
+    //
+    // Note: Legacy errors are present in Case 1 and Case 2. Therefore, newer errors are possibly
+    // encountered in Case 3s, the unnamed field case.
+    let (module_variant_is_struct, module_legacy_err) = if let TypeDef::Variant(details) =
+        dispatch_error_def
+    {
         let module_variant = details
             .variants()
             .iter()
             .find(|variant| variant.name() == "Module")
             .unwrap_or_else(|| {
                 abort_call_site!("DispatchError::Module variant expected in metadata")
             });
-        let are_fields_named = module_variant
-            .fields()
-            .get(0)
-            .unwrap_or_else(|| {
-                abort_call_site!(
-                    "DispatchError::Module expected to contain 1 or more fields"
-                )
-            })
-            .name()
-            .is_some();
-        are_fields_named
+
+        let module_field = module_variant.fields().get(0).unwrap_or_else(|| {
```

Comment:
I find it a little hard following the two-booleans and nested if chains; how about defining a small enum like

```rust
enum ModuleErrorType {
    Case1,
    Case2,
    Case3,
}
```

And then perhaps tweaking the if statements to get rid of some nesting. Perhaps we can move this check to a different function for instance so that early returns are easy? Something like:

```rust
fn module_error_type(...) -> ModuleErrorType {    
    // if the fields are named, this is case 1; the oldest.
    if !module_field.name().is_none() {
        return ModuleErrorType::Case1
    }

    // get the error struct
    let module_err = metadata
        .types
        .resolve(module_field.ty().id())
        .unwrap_or_else(|| {
            abort_call_site!("sp_runtime::ModuleError type expected in metadata")
        });

    let error_type_def = match module_err.type_def() {
        TypeDef::Composite(composite) => composite,
        _ => abort_call_site!("sp_runtime::ModuleError type should be a composite type")
    };

    let error_field_in_composite = error_type_def
        .fields()
        .iter()
        .find(|field| field.name() == Some(&"error".to_string()))
        .unwrap_or_else(|| {
            abort_call_site!(
                "sp_runtime::ModuleError expected to contain error field"
            )
        });

    // if error type is u8, case2, else assume it's [u8; 4]
    if error_field.type_name() == Some(&"u8".to_string()) {
        ModuleErrorType::Case2
    } else {
        ModuleErrorType::Case3
    }
}
```

(Code completely untested). What do you reckon?

Reply:
Indeed, that makes more sense to have a different function. Especially for supporting multiple versions, better change it now and simplify maintainability going forward :D Thanks for feedback